### ISSUE TITLE
Error in chewbbaca: Uniprot annotation step

### ISSUE THREAD

**Issue by bala-ruokavirasto** [B-UMMI/chewBBACA#126]
Dear Chewbbaca Team,

I am getting the below error when i try to annotate the schema using the below commands:

#### Chewbbaca version used ####
docker pull quay.io/biocontainers/chewbbaca:2.8.5--pyhdfd78af_0
###########################

### Commands to replicate this error ####
#curl -o datasets 'https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets'
#./datasets download genome taxon "yersinia enterocolitica"
#cp ncbi_dataset/data/*/GC*_genomic.fna yenterocolitica_complete_genomes/.
singularity build chewbbaca.sif docker://quay.io/biocontainers/chewbbaca:2.8.5--pyhdfd78af_0
singularity  exec -B $PWD chewbbaca.sif prodigal -a prodigal_inputs/GCF_001160345.1_8016_2_61_protein.faa -i prodigal_inputs/GCF_001160345.1_8016_2_61_genomic.fna -t Yersinia_enterocolitica_training_file.trn
singularity  exec -B $PWD chewbbaca.sif chewBBACA.py CreateSchema -i yenterocolitica_complete_genomes -o Y.enterocolitica_schema_18052022 --ptf Yersinia_enterocolitica_training_file.trn --cpu 39
singularity  exec -B $PWD chewbbaca.sif chewBBACA.py UniprotFinder -i Y.enterocolitica_schema_18052022 -o annotations_test_180522 --cpu 39
################################

###### Error message below #####

chewBBACA version: 2.8.5
Authors: Mickael Silva, Pedro Cerqueira, Rafael Mamede
Github: https://github.com/B-UMMI/chewBBACA
Wiki: https://github.com/B-UMMI/chewBBACA/wiki
Tutorial: https://github.com/B-UMMI/chewBBACA_tutorial
Contacts: imm-bioinfo@medicina.ulisboa.pt

=============================
  chewBBACA - UniprotFinder
=============================
Started at: 2022-05-18T17:08:44

Schema: test_compatible_180522
Number of loci: 16503

Querying UniProt's SPARQL endpoint...

**Error on get_annotation:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/utils/multiprocessing_operations.py", line 39, in function_helper
    results = input_args[-1](*input_args[0:-1])
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/utils/uniprot_requests.py", line 246, in get_annotation
    name, url, label = select_name(result)
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/utils/uniprot_requests.py", line 70, in select_name
    current_res = aux[i]
IndexError: list index out of range

Error on get_annotation:
Traceback (most recent call last):
 [====================] 100%Traceback (most recent call last):/multiprocessing_operations.py", line 39, in function
  File "/usr/local/bin/chewBBACA.py", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/chewBBACA.py", line 1480, in main
    functions_info[process][1]()
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/utils/process_datetime.py", line 149, in wrapper
    func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/chewBBACA.py", line 1076, in find_uniprot
    uniprot_find.main(**vars(args))
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/UniprotFinder/uniprot_find.py", line 527, in main
    output_table = create_annotations_table(annotations, output_directory,
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/UniprotFinder/uniprot_find.py", line 418, in create_annotations_table
    new_lines = ['\t'.join(l) for l in new_lines]
  File "/usr/local/lib/python3.9/site-packages/CHEWBBACA/UniprotFinder/uniprot_find.py", line 418, in <listcomp>
    new_lines = ['\t'.join(l) for l in new_lines]
TypeError: sequence item 2: expected str instance, list found**

Hope i can able to solve this issue and i tried many ways before posting here such as using Prepexternalschema, etc... None of the approaches help. I think the problem is with some bug in the software.

Please clarify!

Best Regards,
Bala

**Comment by rfm-targa** [comment]
Hello @bala-ruokavirasto,

Thank you for reporting this issue. We have found the cause of this bug and will include a bugfix in the next PR. The next PR will include a lot of changes and must be reviewed very carefully. I will let you know when it is merged into the master branch.

Best regards,

Rafael

**Comment by bala-ruokavirasto** [comment]
Thanks a lot! It will be very useful if fixed as I need to annotate some of my custom-made schemas for different projects!

Best Regards,
Bala

**Comment by tanzhizhou** [comment]
I have met the same problem today:
```python
Error on get_annotation:
Traceback (most recent call last):
  File "/home/tanzhizhou/Miniconda3/envs/chewie/lib/python3.7/site-packages/CHEWBBACA/utils/multiprocessing_operations.py", line 39, in function_helper
    results = input_args[-1](*input_args[0:-1])
  File "/home/tanzhizhou/Miniconda3/envs/chewie/lib/python3.7/site-packages/CHEWBBACA/utils/uniprot_requests.py", line 246, in get_annotation
    name, url, label = select_name(result)
  File "/home/tanzhizhou/Miniconda3/envs/chewie/lib/python3.7/site-packages/CHEWBBACA/utils/uniprot_requests.py", line 70, in select_name
    current_res = aux[i]
IndexError: list index out of range
```

I really love chewbacaa, hope this bug could be resolved soon!

**Comment by rfm-targa** [comment]
Hello @tanzhizhou,

Thank you for your interest in chewBBACA. A fix for this issue was added to the main branch with the last PR. The PR adds a lot of changes and will be a major release (from v2.8.5 to v3.0.0). You can expect a new release that you can install with Pip or Conda by the end of the month.

Best regards,

Rafael

**Comment by tanzhizhou** [comment]
Hi Rafael,
Many thanks for your detailed reply! chewBBACA v3.0, awesome!

All the best,

Zhizhou
China CDC

**Comment by ramirma** [comment]
Dear Zhizhou,

Thank you for your interest in chewBBACA and I am glad you liked 3.0. Just a word of caution that we do not consider its current version final and there may be a few bugs to squash. If you find one please let us know.

Best Regards,

Mario

**Comment by rfm-targa** [comment]
@bala-ruokavirasto and @tanzhizhou, I've created a [release](https://github.com/B-UMMI/chewBBACA/releases/tag/v3.0.0) for chewBBACA 3.0. You can install v3.0.0 with [pip](https://pypi.org/project/chewBBACA/#description) and it'll be available through conda soon. Let us know if you try to run the UniprotFinder module. The bugs should be fixed, but your feedback is important to know it works well outside our test setup.

**Comment by tanzhizhou** [comment]
> 

That is a really great news! Many thanks Rafael! I will use it ASAP. I wish you all the best for the coming Christmas.

Zhizhou

**Comment by rfm-targa** [comment]
Happy holidays @tanzhizhou.

**Comment by rfm-targa** [comment]
Greetings! The link to the UniProt FTP has changed, and the UniprotFinder module was failing to retrieve the reference proteomes when the `--taxa` parameter was provided. We have just released chewBBACA v3.2.0, which includes a bug fix for this issue. The link has been updated, and the UniprotFinder module should work as expected.
I'm closing the issue, but let us know if you have any feedback about chewBBACA v3.2.0 or find any issues.
